app/src/main/java/org/oppia/android/app/home/HomeViewModel.kt

```diff
@@ -0,0 +1,185 @@
+      }
+      return@map itemListResult.getOrDefault(listOf())
+    }
+  }
+
+  /**
+   * Returns a [HomeItemViewModel] corresponding to the welcome message (see [WelcomeViewModel]), or null if
+   * the specified profile has insufficient information to show the welcome message.
+   * */
+  private fun computeWelcomeViewModel(profile: Profile): HomeItemViewModel? {
+    return if (profile.name.isNotEmpty()) {
+      WelcomeViewModel(fragment, oppiaClock, profile.name)
+    } else null
+  }
+
+  /**
+   * Returns a [HomeItemViewModel] corresponding to the promoted stories to be displayed for this learner
+   * (see [PromotedStoryListViewModel]), or null if this profile does not have any promoted stories.
+   * Promoted stories are determined by any recent stories started by this profile.
+   * */
+  private fun computePromotedStoryListViewModel(
+    ongoingStoryList: OngoingStoryList
+  ): HomeItemViewModel? {
+    val storyViewModelList = computePromotedStoryViewModelList(ongoingStoryList)
+    return if (storyViewModelList.isNotEmpty()) {
+      return PromotedStoryListViewModel(
+        activity,
+        internalProfileId,
+        intentFactoryShim,
+        storyViewModelList
+      )
+    } else null
+  }
+
+  /**
+   * Returns a list of [HomeItemViewModel]s corresponding to the the [PromotedStoryListViewModel] displayed
+   * for this profile (see [PromotedStoryViewModel]), or an empty list if the profile does not have any
+   * ongoing stories at all.
+   */
+  private fun computePromotedStoryViewModelList(
+    ongoingStoryList: OngoingStoryList
+  ): List<PromotedStoryViewModel> {
+    val storyList = if (ongoingStoryList.recentStoryCount != 0) {
+      ongoingStoryList.recentStoryList
+    } else {
+      // TODO(#936): Optimise this as part of recommended stories.
+      ongoingStoryList.olderStoryList
+    }
+    return storyList.take(promotedStoryListLimit)
+      .map { promotedStory ->
+        PromotedStoryViewModel(
+          activity,
+          internalProfileId,
+          intentFactoryShim,
+          storyList.size,
+          storyEntityType,
+          promotedStory
+        )
+      }
+  }
+
+  /**
+   * Returns a list of [HomeItemViewModel]s corresponding to all the lesson topics available and to be
+   * displayed on the home activity (see [TopicSummaryViewModel]) along with associated topics list header (see
+   * [AllTopicsViewModel]). Returns null if there are no lesson topics to display in the home fragment.
+   */
+  private fun computeAllTopicsItemsViewModelList(
+    topicList: TopicList
+  ): Iterable<HomeItemViewModel>? {
+    val allTopicsList = topicList.topicSummaryList.mapIndexed { topicIndex, topicSummary ->
+      TopicSummaryViewModel(
+        activity,
+        topicSummary,
+        topicEntityType,
+        fragment as TopicSummaryClickListener,
+        position = topicIndex
+      )
+    }
+    return if (!allTopicsList.isEmpty()) {
+      listOf(AllTopicsViewModel()) + allTopicsList
```

Comment:
Does this op work without the listOf() bit?

Reply:
It gives `Unexpected tokens (use ';' to separate expressions on the same line)` error, and since allTopicsList isn't declared as mutable using any sort of `.add` isn't possible either.

If we changed it to a MutableList then this works, but I personally prefer using listOf():
`val allTopicsList = topicList.topicSummaryList.mapIndexed { topicIndex, topicSummary ->
      TopicSummaryViewModel(
        activity,
        topicSummary,
        topicEntityType,
        fragment as TopicSummaryClickListener,
        position = topicIndex
      )
    }.toMutableList()
    if (!allTopicsList.isEmpty()) {
      val list = allTopicsList as MutableList<HomeItemViewModel>
      list.add(0, AllTopicsViewModel())
      return list
    } else return emptyList()`

Reply:
Ah, I was hoping there was an extension function in Kotlin defined like:

```kotlin
fun <T> T.plus(list: List<T>): List<T> { ... }
```

For element + list concatenation (note that list + element concatenation does work and doesn't require mutability since it creates a new list). Unfortunately this particular order doesn't work. :(